<?xml version="1.0"?>
<launch>
    <!-- Let's use sim time! -->
    <param name="/use_sim_time" value="true"/>

    <!-- arguments -->
    <arg name="nodelet_manager" default="velodyne_nodelet_manager" />
    <arg name="points_topic" default="/lidar_points" />
    <arg name="imu_topic" default="/imu/data" />
    <arg name="gps_topic" default="/fix" />
    <arg name="base_link" default="base_link" />
    <arg name="odom_child_frame_id" default="base_link" />
    <arg name="map_frame_id" default="world" />
    <arg name="maps" default="[/share/itri/map.pcd]" />
    <arg name="map_resolution" default="2.0" />
    <arg name="output" default="/share/eval_csv_" />
    <arg name="ground_truth" default="/share/itri/ITRI_Public_Ground_Truth.csv" />

    <!-- in case you use velodyne_driver, comment out the following line -->
    <node pkg="nodelet" type="nodelet" name="$(arg nodelet_manager)" args="manager" output="screen"/>

    <!-- initializer node -->
    <node 
        pkg="tools"
        name="Initializer"
        type="initializer.py"
        output="screen"
        args="-g $(arg gps_topic) -m '/globalmap' -s /score -f $(arg map_frame_id) -r 40"
    />

    <!-- hdl_localization_nodelet -->
    <node pkg="nodelet" type="nodelet" name="hdl_localization_nodelet" args="load hdl_localization/HdlLocalizationNodelet $(arg nodelet_manager)">
        <remap from="/velodyne_points" to="$(arg points_topic)" />
        <remap from="/initialpose" to="Initializer/init" />
        <!--<remap from="/guesses" to="Initializer/guess" />-->
        <!--<remap from="/globalmap" to="Initializer/submap" />-->
        <!-- odometry frame_id -->
        <param name="map_frame_id" value="$(arg map_frame_id)" />
        <param name="odom_child_frame_id" value="$(arg odom_child_frame_id)" />
        <!-- imu settings -->
        <!-- during "cool_time", imu inputs are ignored -->
        <param name="use_guesses" value="true" />
        <param name="cool_time_duration" value="1.0" />
        <!-- ndt settings -->
        <!-- if NDT is slow for your PC, try DIRECT1 serach method, which is a bit unstable but extremely fast -->
        <param name="ndt_neighbor_search_method" value="DIRECT7" />
        <param name="ndt_resolution" value="1.0" />
        <param name="downsample_resolution" value="0.5" />
        <!-- if "specify_init_pose" is true, pose estimator will be initialized with the following params -->
        <!-- otherwise, you need to input an initial pose with "2D Pose Estimate" on rviz" -->
        <param name="specify_init_pose" value="false" />
    </node>
    
    <!-- Run Rviz-->
    <node pkg="rviz" type="rviz" name="rviz" args="-d $(find hdl_localization)/rviz/itri.rviz" />

    <!-- globalmap_server_nodelet -->
    <node pkg="nodelet" type="nodelet" name="globalmap_server_nodelet" args="load hdl_localization/GlobalmapServerNodelet $(arg nodelet_manager)">
        <rosparam param="globalmaps_pcd" subst_value="True">$(arg maps)</rosparam>
        <param name="map_frame" value="$(arg map_frame_id)" />
        <param name="downsample_resolution" value="$(arg map_resolution)" />
    </node>

    <!-- CSV Recorder -->
    <node 
        pkg="tools"
        name="CSVRecorder"
        type="ros2csv.py"
        output="screen"
        args="-t '/odom' -o $(arg output)"
    />
</launch>
